crates/ironrdp-acceptor/src/lib.rs

```diff
@@ -0,0 +1,129 @@
+#[macro_use]
+extern crate tracing;
+
+use std::net::SocketAddr;
+
+use ironrdp_pdu::nego;
+use tokio_rustls::TlsAcceptor;
+
+use std::io;
+
+use ironrdp_connector::{custom_err, ConnectorResult, Sequence, Written};
+use ironrdp_tokio::{Framed, FramedRead, FramedWrite, TokioFramed};
+use tokio::io::{AsyncRead, AsyncWrite};
+use tokio_rustls::server::TlsStream;
+
+use self::connection::AcceptorResult;
+
+pub mod channel_connection;
+pub mod connection;
+pub mod finalization;
+pub mod util;
+
+pub use self::connection::ServerAcceptor;
+pub use ironrdp_connector::DesktopSize;
+
+#[derive(Clone)]
+pub struct RdpServerOptions {
+    pub addr: SocketAddr,
+    pub security: RdpServerSecurity,
+}
+
+#[derive(Clone)]
+pub enum RdpServerSecurity {
+    None,
+    SSL(TlsAcceptor),
+}
+
+impl RdpServerSecurity {
+    pub fn flag(&self) -> nego::SecurityProtocol {
+        match self {
+            RdpServerSecurity::None => ironrdp_pdu::nego::SecurityProtocol::empty(),
+            RdpServerSecurity::SSL(_) => ironrdp_pdu::nego::SecurityProtocol::SSL,
+        }
+    }
+}
+
+pub enum BeginResult<S> {
+    ShouldUpgrade(S),
+    Continue(TokioFramed<S>),
+}
+
+pub async fn accept_begin<S>(stream: S, acceptor: &mut ServerAcceptor) -> ConnectorResult<BeginResult<S>>
+where
+    S: Unpin + AsyncRead + AsyncWrite,
+{
+    let mut buf = Vec::new();
+    let mut framed = TokioFramed::new(stream);
+
+    loop {
+        if let Some(security) = acceptor.reached_security_upgrade() {
+            let result = if security.is_empty() {
+                BeginResult::Continue(framed)
+            } else {
+                BeginResult::ShouldUpgrade(framed.into_inner().0)
```

Comment:
For sanity, `.into_inner_no_leftover()` should be used instead. This function [asserts that there is no leftover bytes](https://github.com/Devolutions/IronRDP/blob/55176ff99b8c72076c49ec4a2333ba9324798bc5/crates/ironrdp-async/src/framed.rs#L66) before proceeding further. Also, I’m not convinced that this logic should lives here. I would suggest an [API similar to the one we have for the connector](https://github.com/Devolutions/IronRDP/blob/55176ff99b8c72076c49ec4a2333ba9324798bc5/crates/ironrdp-client/src/rdp.rs#L97-L109), so it’s easier to support an alternative to tokio in the future.